### PR TITLE
DEV: Add summarization logic from core

### PR DESCRIPTION
**Pull request by keegangeorge**
This PR adds summarization logic back to the plugin from core. See core PR: https://github.com/discourse/discourse/pull/27373

### CODE REVIEW

File: db/migrate/20240606152117_copy_summary_sections_to_ai_summaries.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+class CopySummarySectionsToAiSummaries < ActiveRecord::Migration[7.0]
```
Comment (romanrizzi) [inline db/migrate/20240606152117_copy_summary_sections_to_ai_summaries.rb:3]: This should be under `post_migrate`, and it should be irreversible.

File: db/migrate/20240606151348_create_ai_summaries_table.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+class CreateAiSummariesTable < ActiveRecord::Migration[7.0]
+  def change
+    create_table :ai_summaries do |t|
+      t.integer :target_id, null: false
+      t.string :target_type, null: false
+      t.int4range :content_range
+      t.string :summarized_text, null: false
+      t.integer :meta_section_id
```
Comment (romanrizzi) [inline db/migrate/20240606151348_create_ai_summaries_table.rb:10]: We should drop this, and only migrate those where `meta_section_id` is null. This means they are the final summary shown to users. We never really use these small fragments so it's a good opportunity to get rid of them

File: config/settings.yml [hunk 3]
```diff
@@ -317,6 +317,15 @@ discourse_ai:
   ai_summarization_discourse_service_api_key:
     default: ""
     secret: true
+  ai_summarization_strategy:
```
Comment (romanrizzi) [inline config/settings.yml:320]: I think we are missing a way to migrate the old setting here? We don't want to break summaries

File: app/services/topic_summarization.rb [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    main_summary =
+      AiSummary.create!(
+        target: topic,
+        algorithm: strategy.display_name,
+        content_range: (post_numbers.first..post_numbers.last),
+        summarized_text: result[:summary],
+        original_content_sha: build_sha(post_numbers),
+      )
+
+    result[:chunks].each do |chunk|
```
Comment (romanrizzi) [inline app/services/topic_summarization.rb:112]: Dropping `meta_section_id` means we can simplify this

**Review by romanrizzi — CHANGES_REQUESTED** [review]

File: app/services/topic_summarization.rb (hunk 4 above)
Comment (romanrizzi) [reply app/services/topic_summarization.rb:112]: Yes, exactly

File: db/post_migrate/20240610232546_copy_custom_summarization_allowed_groups_to_ai_custom_summarization_allowed_groups.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+class CopyCustomSummarizationAllowedGroupsToAiCustomSummarizationAllowedGroups < ActiveRecord::Migration[
+  7.0
```
Comment (xfalcox) [inline db/post_migrate/20240610232546_copy_custom_summarization_allowed_groups_to_ai_custom_summarization_allowed_groups.rb:4]: lol

File: lib/summarization/entry_point.rb [hunk 6]
```diff
@@ -67,7 +81,10 @@ def inject_into(plugin)
         #end
 
         foldable_models.each do |model|
-          plugin.register_summarization_strategy(Strategies::FoldContent.new(model))
+          DiscoursePluginRegistry.register_summarization_strategy(
```
Comment (romanrizzi) [inline lib/summarization/entry_point.rb:84]: I think we can just return `foldable_models` in `Models::Base#available_strategies`? Nobody else consumes this register

**Review by romanrizzi — CHANGES_REQUESTED** [review]

**Review by romanrizzi — APPROVED** [review]

**Pull request merged**